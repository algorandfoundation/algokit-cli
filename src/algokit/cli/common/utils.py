# Cli/Click related helper functions and classes


import typing as t

import click

from algokit.cli.common.constants import ExplorerEntityType


class MutuallyExclusiveOption(click.Option):
    """
    A Click option that defines mutually exclusive command line options.

    Args:
        *args: Positional arguments passed to the parent class constructor.
        **kwargs: Keyword arguments passed to the parent class constructor.
            not_required_if (list): A list of options that the current option is mutually exclusive with.

    Attributes:
        not_required_if (list): A list of options that the current option is mutually exclusive with.

    Raises:
        AssertionError: If the `not_required_if` parameter is not provided.

    Example:
        ```python
        @click.command()
        @click.option('--option1', help='Option 1')
        @click.option('--option2', help='Option 2')
        @click.option('--option3', help='Option 3', cls=MutuallyExclusiveOption, not_required_if=['option1', 'option2'])
        def my_command(option1, option2, option3):
            # Command logic here
            pass
        ```

        In the example above, the `MutuallyExclusiveOption` class is used to define the `option3` command line option.
        This option is mutually exclusive with `option1` and `option2`,
        meaning that if either `option1` or `option2` is provided, `option3` cannot be used.
        If `option3` is provided along with `option1` or `option2`, a `click.UsageError` is raised.
    """

    def __init__(self, *args: t.Any, **kwargs: t.Any) -> None:
        self.not_required_if: list = kwargs.pop("not_required_if")

        assert self.not_required_if, "'not_required_if' parameter required"
        kwargs["help"] = (
            kwargs.get("help", "") + " Option is mutually exclusive with " + ", ".join(self.not_required_if) + "."
        ).strip()
        super().__init__(*args, **kwargs)

    def handle_parse_result(
        self, ctx: click.Context, opts: t.Mapping[str, t.Any], args: list[str]
    ) -> tuple[t.Any, list[str]]:
        """
        Overrides the `handle_parse_result` method of the `click.Option` class.

        This method checks if the current option is present in the provided options (`opts`)
        and if any of the mutually exclusive options are also present.
        If both the current option and a mutually exclusive option are present, it raises a `click.UsageError`.
        Otherwise, it returns the result of the parent `handle_parse_result` method.

        Args:
            ctx (click.Context): The Click context object.
            opts (dict): The dictionary of parsed options.
            args (list): The list of remaining arguments.

        Returns:
            The result of the parent `handle_parse_result` method.

        Raises:
            click.UsageError: If the current option and a mutually exclusive option are both present.
        """

        current_opt: bool = self.name in opts
        for mutex_opt in self.not_required_if:
            if mutex_opt in opts:
                if current_opt:
                    raise click.UsageError(
                        "Illegal usage: '" + str(self.name) + "' is mutually exclusive with " + str(mutex_opt) + "."
                    )
                self.prompt = None
        return super().handle_parse_result(ctx, opts, args)


def get_explorer_url(identifier: str | int, network: str, entity_type: ExplorerEntityType) -> str:
    """
    Returns a URL for exploring a specified type (transaction, asset, address) on the specified network.

    Args:
        identifier (str | int): The ID of the transaction, asset, or address to explore.
        network (str): The name of the network (e.g., "localnet", "testnet", "mainnet").
        entity_type (ExplorerEntityType): The type to explore (e.g., ExplorerEntityType.TRANSACTION,
        ExplorerEntityType.ASSET, ExplorerEntityType.ADDRESS).

    Returns:
        str: The URL for exploring the specified type on the specified network.

    Raises:
        ValueError: If the network or explorer type is invalid.
    """

    base_urls: dict[str, dict[str, str]] = {
        "testnet": {
            ExplorerEntityType.TRANSACTION.value: "https://testnet.algoexplorer.io/tx/",
            ExplorerEntityType.ASSET.value: "https://testnet.explorer.perawallet.app/assets/",
            ExplorerEntityType.ADDRESS.value: "https://testnet.algoexplorer.io/address/",
        },
        "mainnet": {
            ExplorerEntityType.TRANSACTION.value: "https://algoexplorer.io/tx/",
            ExplorerEntityType.ASSET.value: "https://explorer.perawallet.app/assets/",
            ExplorerEntityType.ADDRESS.value: "https://algoexplorer.io/address/",
        },
    }

    if network == "localnet":
        return f"https://app.dappflow.org/setnetwork?name=sandbox&redirect=explorer/{entity_type.value}/{identifier}/"

    if network not in base_urls:
        raise ValueError(f"Invalid network: {network}")

    if entity_type.value not in base_urls[network]:
        raise ValueError(f"Invalid explorer type: {entity_type}")

    return base_urls[network][entity_type.value] + str(identifier)
