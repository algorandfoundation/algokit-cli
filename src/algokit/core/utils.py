import re
import socket


def extract_version_triple(version_str: str) -> str:
    match = re.search(r"\d+\.\d+\.\d+", version_str)
    if not match:
        raise ValueError("Unable to parse version number")
    return match.group()


def is_minimum_version(system_version: str, minimum_version: str) -> bool:
    system_version_as_tuple = tuple(map(int, system_version.split(".")))
    minimum_version_as_tuple = tuple(map(int, minimum_version.split(".")))
    return system_version_as_tuple >= minimum_version_as_tuple


def is_network_available(host: str = "8.8.8.8", port: int = 53, timeout: float = 3.0) -> bool:
    """
    Check if internet is available by trying to establish a socket connection.
    """

    try:
        socket.setdefaulttimeout(timeout)
        socket.socket(socket.AF_INET, socket.SOCK_STREAM).connect((host, port))
        return True
    except OSError:
        return False
