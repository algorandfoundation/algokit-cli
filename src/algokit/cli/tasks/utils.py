# AlgoKit Tasks related utility functions

import logging
import os
import stat
import sys

import algosdk
import algosdk.encoding
import click
from algokit_utils import (
    Account,
    get_algod_client,
    get_algonode_config,
    get_default_localnet_config,
)
from algosdk.util import algos_to_microalgos, microalgos_to_algos

from algokit.cli.common.constants import AlgorandNetwork
from algokit.core.tasks.wallet import get_alias

logger = logging.getLogger(__name__)


def _validate_asset_balance(account_info: dict, asset_id: int, decimals: int, amount: int = 0) -> None:
    asset_record = next((asset for asset in account_info.get("assets", []) if asset["asset-id"] == asset_id), None)

    if not asset_record:
        raise click.ClickException("Account is not opted into the asset")

    if amount > 0 and asset_record["amount"] < amount:
        required = amount / 10**decimals
        available = asset_record["amount"] / 10**decimals
        raise click.ClickException(
            f"Insufficient asset balance in account, required: {required}, available: {available}"
        )


def _validate_algo_balance(account_info: dict, amount: int) -> None:
    if account_info.get("amount", 0) < amount:
        required = microalgos_to_algos(amount)  # type: ignore[no-untyped-call]
        available = microalgos_to_algos(account_info.get("amount", 0))  # type: ignore[no-untyped-call]
        raise click.ClickException(
            f"Insufficient Algos balance in account, required: {required} Algos, available: {available} Algos"
        )


def get_private_key_from_mnemonic() -> str:
    """
    Converts a mnemonic phrase into a private key.

    Returns:
        str: The private key generated from the mnemonic phrase.

    Raises:
        click.ClickException: If the entered mnemonic phrase is invalid.

    Example Usage:
        private_key = get_private_key_from_mnemonic()
        print(private_key)

    Inputs:
        None

    Flow:
        1. Prompts the user to enter a mnemonic phrase.
        2. Converts the entered mnemonic phrase into a private key using `algosdk.mnemonic.to_private_key`.
        3. Returns the private key.

    """

    mnemonic_phrase = click.prompt("Enter the mnemonic phrase (25 words separated by whitespace)", hide_input=True)
    try:
        return str(algosdk.mnemonic.to_private_key(mnemonic_phrase))  # type: ignore[no-untyped-call]
    except Exception as err:
        raise click.ClickException("Invalid mnemonic. Please provide a valid Algorand mnemonic.") from err


def load_algod_client(network: AlgorandNetwork) -> algosdk.v2client.algod.AlgodClient:
    """
    Returns an instance of the `algosdk.v2client.algod.AlgodClient` class for the specified network.

    Args:
        network (str): The network for which the `AlgodClient` instance needs to be loaded.

    Returns:
        algosdk.v2client.algod.AlgodClient: An instance of the `AlgodClient` class for the specified network.

    Raises:
        click.ClickException: If the specified network is invalid.
    """

    config_mapping = {
        AlgorandNetwork.LOCALNET: get_default_localnet_config("algod"),
        AlgorandNetwork.TESTNET: get_algonode_config("testnet", "algod", ""),
        AlgorandNetwork.MAINNET: get_algonode_config("mainnet", "algod", ""),
    }
    try:
        return get_algod_client(config_mapping[network])
    except KeyError as err:
        raise click.ClickException("Invalid network") from err


def get_asset_decimals(asset_id: int, algod_client: algosdk.v2client.algod.AlgodClient) -> int:
    """
    Retrieves the number of decimal places for a given asset.

    Args:
        asset_id (int): The ID of the asset for which the number of decimal places is to be retrieved. (0 for Algo)
        algod_client (algosdk.v2client.algod.AlgodClient): An instance of the AlgodClient class.

    Returns:
        int: The number of decimal places for the given asset.

    Raises:
        click.ClickException: If the asset info response is invalid.

    Example:
        asset_id = 123
        algod_client = algosdk.v2client.algod.AlgodClient("https://api.algoexplorer.io", "API_KEY")
        decimals = get_asset_decimals(asset_id, algod_client)
        print(decimals)
    """

    if asset_id == 0:
        return 6

    asset_info = algod_client.asset_info(asset_id)

    if not isinstance(asset_info, dict) or "params" not in asset_info or "decimals" not in asset_info["params"]:
        raise click.ClickException("Invalid asset info response")

    return int(asset_info["params"]["decimals"])


def validate_balance(
    algod_client: algosdk.v2client.algod.AlgodClient, account: Account | str, asset_id: int, amount: int = 0
) -> None:
    """
    Validates the balance of an account before an operation.

    Args:
        algod_client (algosdk.v2client.algod.AlgodClient): The AlgodClient object for
        interacting with the Algorand blockchain.
        account (Account | str): The account object.
        asset_id (int): The ID of the asset to be checked (0 for Algos).
        amount (int): The amount of Algos or asset for the operation. Defaults to 0 implying opt-in check only.

    Raises:
        click.ClickException: If any validation check fails.
    """
    address = account.address if isinstance(account, Account) else account
    account_info = algod_client.account_info(address)

    if not isinstance(account_info, dict):
        raise click.ClickException("Invalid account info response")

    if asset_id == 0:
        _validate_algo_balance(account_info, amount)
    else:
        decimals = get_asset_decimals(asset_id, algod_client)
        _validate_asset_balance(account_info, asset_id, decimals, amount)


def validate_address(address: str) -> None:
    """
    Check if a given address is a valid Algorand account address.

    Args:
        address (str): The address to be validated.

    Raises:
        click.ClickException: If the address is invalid.

    Returns:
        None
    """

    if not algosdk.encoding.is_valid_address(address):  # type: ignore[no-untyped-call]
        raise click.ClickException(f"`{address}` is an invalid account address")


def get_account_with_private_key(address: str) -> Account:
    """
    Retrieves an account object with the private key based on the provided address.

    Args:
        address (str): The address for which to retrieve the account object.

    Returns:
        Account: An account object with the address and private key.

    Raises:
        click.ClickException: If the address is not valid or if the alias does not exist or does not have a private key.
    """

    parsed_address = address.strip('"')

    try:
        validate_address(parsed_address)
        pk = get_private_key_from_mnemonic()
        return Account(address=parsed_address, private_key=pk)
    except click.ClickException as ex:
        alias_data = get_alias(parsed_address)

        if not alias_data:
            raise click.ClickException(f"Alias `{parsed_address}` alias does not exist.") from ex
        if not alias_data.private_key:
            raise click.ClickException(f"Alias `{parsed_address}` does not have a private key.") from ex

        return Account(address=alias_data.address, private_key=alias_data.private_key)


def get_address(address: str) -> str:
    """
    Validates the given address and returns it if valid. If the address is not valid,
    it checks if the address is an alias and returns the corresponding address if the alias exists.

    Args:
        address (str): The address to be validated or checked for an alias.

    Returns:
        str: The validated address or the corresponding address from an alias.

    Raises:
        click.ClickException: If the address is not valid and no corresponding alias exists.

    Example:
        address = get_address("ABCD1234")
        print(address)
    """

    parsed_address = address.strip('"')

    try:
        validate_address(parsed_address)
        return parsed_address
    except click.ClickException as ex:
        if len(parsed_address) == algosdk.constants.address_len:
            raise click.ClickException(f"`{parsed_address}` is an invalid account address") from ex

        alias_data = get_alias(parsed_address)

        if not alias_data:
            raise click.ClickException(f"Alias `{parsed_address}` alias does not exist.") from ex

        return alias_data.address


def stdin_has_content() -> bool:
    """
    Checks if there is content in the standard input.

    Returns:
        bool: True if there is content in the standard input, False otherwise.
    """

    mode = os.fstat(sys.stdin.fileno()).st_mode
    return stat.S_ISFIFO(mode) or stat.S_ISREG(mode)


def validate_account_balance_to_opt_in(
    algod_client: algosdk.v2client.algod.AlgodClient, account: Account, num_assets: int
) -> None:
    """
    Validates the balance of an account before opt in operation.
    Each asset requires 0.1 Algos to opt in.

    Args:
        algod_client (algosdk.v2client.algod.AlgodClient): The AlgodClient object for
        interacting with the Algorand blockchain.
        account (Account | str): The account object.
        num_assets (int): The number of the assets for opt in (0 for Algos).

    Raises:
        click.ClickException: If there is an insufficient fund in the account or account is not valid.
    """

    address = account.address if isinstance(account, Account) else account
    account_info = algod_client.account_info(address)

    if not isinstance(account_info, dict):
        raise click.ClickException("Invalid account info response")

    required_microalgos = num_assets * algos_to_microalgos(0.1)  # type: ignore[no-untyped-call]
    available_microalgos = account_info.get("amount", 0)
    if available_microalgos < required_microalgos:
        required_algo = microalgos_to_algos(required_microalgos)  # type: ignore[no-untyped-call]
        available_algos = microalgos_to_algos(available_microalgos)  # type: ignore[no-untyped-call]
        raise click.ClickException(
            f"Insufficient Algos balance in account to opt in, required: {required_algo} Algos, available:"
            f" {available_algos} Algos"
        )


def get_account_info(algod_client: algosdk.v2client.algod.AlgodClient, account_address: str) -> dict:
    account_info = algod_client.account_info(account_address)
    assert isinstance(account_info, dict)
    return account_info
