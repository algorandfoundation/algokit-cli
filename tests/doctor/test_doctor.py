import typing
from datetime import datetime
from pathlib import Path

import click
import pytest
from approvaltests.pytest.py_test_namer import PyTestNamer
from approvaltests.scrubbers.scrubbers import Scrubber
from pytest_mock import MockerFixture

from tests.utils.approvals import TokenScrubber, combine_scrubbers, verify
from tests.utils.click_invoker import invoke
from tests.utils.proc_mock import ProcMock

PARENT_DIRECTORY = Path(__file__).parent

DOCKER_COMPOSE_VERSION_COMMAND = ["docker", "compose", "version", "--format", "json"]


class VersionInfoType(typing.NamedTuple):
    major: int
    minor: int
    micro: int
    releaselevel: str
    serial: int


@pytest.fixture
def _mock_doctor_dependencies(mocker: MockerFixture) -> None:
    mocker.patch("algokit.cli.doctor.get_current_package_version").return_value = "1.2.3"
    mocker.patch("algokit.cli.doctor.get_latest_github_version").return_value = "1.2.3"
    # Mock datetime
    mocker.patch("algokit.cli.doctor.dt").datetime.now.side_effect = lambda _, tz=None: datetime(
        1990, 12, 31, 10, 9, 8, tzinfo=tz
    )
    # Mock shutil
    mocker.patch("algokit.core.doctor.which").side_effect = mock_shutil_which
    # Mock sys - Tuple[int, int, int, str, int]
    sys_module = mocker.patch("algokit.cli.doctor.sys")
    sys_module.version = "3.6.2"
    sys_module.prefix = "/home/me/.local/pipx/venvs/algokit"
    # Mock enable binary mode to ignore outputting package information to
    # simplify snapshot diffs - otherwise each new run may fail whenever main prod
    # dependencies are updated
    mocker.patch("algokit.cli.doctor.is_binary_mode").return_value = True


@pytest.fixture(autouse=True)
def _mock_happy_values(proc_mock: ProcMock) -> None:
    proc_mock.set_output(["winget", "--version"], ["v1.8.1911"])
    proc_mock.set_output(["brew", "--version"], ["Homebrew 3.6.15", "Homebrew/homebrew-core (blah)"])
    proc_mock.set_output(["docker", "--version"], ["Docker version 20.10.21, build baeda1f"])
    proc_mock.set_output(DOCKER_COMPOSE_VERSION_COMMAND, ['{"version": "v2.12.2"}'])
    proc_mock.set_output(["git", "--version"], ["git version 2.37.1 (Apple Git-137.1)"])
    proc_mock.set_output(["python", "--version"], ["Python 3.10.0"])
    proc_mock.set_output(["python3", "--version"], ["Python 3.11.0"])
    proc_mock.set_output(["pipx", "--version"], ["1.1.0"])
    proc_mock.set_output(["poetry", "--version"], ["blah blah", "", "Poetry (version 1.2.2)"])
    proc_mock.set_output(["node", "--version"], ["v18.12.1"])
    proc_mock.set_output(["npm", "--version"], ["8.19.2"])
    proc_mock.set_output(["npm.cmd", "--version"], ["8.19.2"])


def mock_shutil_which(python_command_name: str) -> str:
    if python_command_name == "python":
        return "/usr/local/bin/python"
    if python_command_name == "python3":
        return "/usr/local/bin/python3"
    return ""


def make_output_scrubber(**extra_tokens: str) -> Scrubber:
    default_tokens = {"test_parent_directory": str(PARENT_DIRECTORY)}
    tokens = default_tokens | extra_tokens
    return combine_scrubbers(
        click.unstyle,
        TokenScrubber(tokens=tokens),
        TokenScrubber(tokens={"test_parent_directory": str(PARENT_DIRECTORY).replace("\\", "/")}),
        lambda t: t.replace("{test_parent_directory}\\", "{test_parent_directory}/"),
    )


@pytest.mark.usefixtures("_mock_doctor_dependencies")
def test_doctor_help() -> None:
    result = invoke("doctor -h")

    assert result.exit_code == 0
    verify(result.output)


@pytest.mark.usefixtures("_mock_doctor_dependencies")
@pytest.mark.mock_platform_system("Darwin")
def test_doctor_with_copy(mocker: MockerFixture) -> None:
    # Mock pyclip
    mocked_os = mocker.patch("algokit.cli.doctor.pyclip.copy")
    result = invoke("doctor -c")

    assert result.exit_code == 0
    mocked_os.assert_called_once()
    verify(result.output, scrubber=make_output_scrubber())


@pytest.mark.usefixtures("_mock_doctor_dependencies", "mock_platform_system")
def test_doctor_successful(request: pytest.FixtureRequest) -> None:
    result = invoke("doctor")

    assert result.exit_code == 0
    verify(result.output, scrubber=make_output_scrubber(), namer=PyTestNamer(request))


@pytest.mark.usefixtures("_mock_doctor_dependencies")
@pytest.mark.mock_platform_system("Darwin")
def test_doctor_with_docker_compose_version_warning(proc_mock: ProcMock) -> None:
    proc_mock.set_output(DOCKER_COMPOSE_VERSION_COMMAND, ['{"version": "v2.1.3"}'])

    result = invoke("doctor")

    assert result.exit_code == 1
    verify(result.output, scrubber=make_output_scrubber())


@pytest.mark.usefixtures("_mock_doctor_dependencies")
@pytest.mark.mock_platform_system("Darwin")
def test_doctor_with_docker_compose_version_gitpod(proc_mock: ProcMock) -> None:
    proc_mock.set_output(DOCKER_COMPOSE_VERSION_COMMAND, ['{"version": "v2.10.0-gitpod.0"}'])

    result = invoke("doctor")

    assert result.exit_code == 0
    verify(result.output, scrubber=make_output_scrubber())


@pytest.mark.usefixtures("_mock_doctor_dependencies")
@pytest.mark.mock_platform_system("Darwin")
def test_doctor_with_docker_compose_version_unparseable(proc_mock: ProcMock) -> None:
    proc_mock.set_output(DOCKER_COMPOSE_VERSION_COMMAND, ['{"version": "TEAPOT"}'])

    result = invoke("doctor")

    assert result.exit_code == 1
    verify(result.output, scrubber=make_output_scrubber())


ALL_COMMANDS = [
    ["brew", "--version"],
    ["docker", "--version"],
    DOCKER_COMPOSE_VERSION_COMMAND,
    ["git", "--version"],
    ["python", "--version"],
    ["python3", "--version"],
    ["pipx", "--version"],
    ["poetry", "--version"],
    ["node", "--version"],
    ["npm", "--version"],
    ["npm.cmd", "--version"],
]


@pytest.mark.usefixtures("_mock_doctor_dependencies", "mock_platform_system")
def test_doctor_all_commands_not_found(request: pytest.FixtureRequest, proc_mock: ProcMock) -> None:
    for cmd in ALL_COMMANDS:
        proc_mock.should_fail_on(cmd[0])

    result = invoke("doctor")

    assert result.exit_code == 1
    verify(result.output, scrubber=make_output_scrubber(), namer=PyTestNamer(request))


@pytest.mark.usefixtures("_mock_doctor_dependencies", "mock_platform_system")
def test_doctor_all_commands_bad_exit(request: pytest.FixtureRequest, proc_mock: ProcMock) -> None:
    for cmd in ALL_COMMANDS:
        proc_mock.should_bad_exit_on(cmd, output=["I AM A TEAPOT"])

    result = invoke("doctor")

    assert result.exit_code == 1
    verify(result.output, scrubber=make_output_scrubber(), namer=PyTestNamer(request))


@pytest.mark.usefixtures("_mock_doctor_dependencies")
@pytest.mark.mock_platform_system("Darwin")
def test_doctor_with_weird_values_on_mac(proc_mock: ProcMock) -> None:
    proc_mock.set_output(["brew", "--version"], ["Homebrew 3.6.15-31-g82d89bb"])

    result = invoke("doctor")

    assert result.exit_code == 0
    verify(result.output, scrubber=make_output_scrubber())


@pytest.mark.usefixtures("_mock_doctor_dependencies")
@pytest.mark.mock_platform_system("Darwin")
def test_unparseable_python_version(proc_mock: ProcMock) -> None:
    proc_mock.set_output(["python", "--version"], ["  ", "1-2-3", "  abc  "])

    result = invoke("doctor")

    assert result.exit_code == 0
    verify(result.output, scrubber=make_output_scrubber())


@pytest.mark.usefixtures("_mock_doctor_dependencies", "proc_mock")
@pytest.mark.mock_platform_system("Darwin")
def test_unexpected_exception_locating_executable(mocker: MockerFixture) -> None:
    def which_throw(_cmd: str) -> None:
        raise RuntimeError("OH NO")

    mocker.patch("algokit.core.doctor.which").side_effect = which_throw

    result = invoke("doctor")

    assert result.exit_code == 0
    verify(result.output, scrubber=make_output_scrubber())


@pytest.mark.usefixtures("_mock_doctor_dependencies")
@pytest.mark.mock_platform_system("Darwin")
def test_npm_permission_denied(proc_mock: ProcMock) -> None:
    proc_mock.should_deny_on(["npm"])

    result = invoke("doctor")

    assert result.exit_code == 1
    verify(result.output, scrubber=make_output_scrubber())


@pytest.mark.usefixtures("_mock_doctor_dependencies")
@pytest.mark.mock_platform_system("Darwin")
def test_new_algokit_version_available(request: pytest.FixtureRequest, mocker: MockerFixture) -> None:
    mocker.patch("algokit.cli.doctor.get_latest_github_version").return_value = "4.5.6"
    result = invoke("doctor")

    assert result.exit_code == 0
    verify(result.output, scrubber=make_output_scrubber(), namer=PyTestNamer(request))


@pytest.mark.usefixtures("_mock_doctor_dependencies")
@pytest.mark.mock_platform_system("Windows")
def test_doctor_with_weird_values_on_windows(proc_mock: ProcMock) -> None:
    proc_mock.set_output(["git", "--version"], ["git version 2.31.0.windows.1"])
    proc_mock.set_output(["winget"], ["v1.8.1911", "Winget v1.8.1911"])
    proc_mock.should_fail_on(["npm"])
    proc_mock.set_output(["npm.cmd", "--version"], [" 16.17.0 "])

    result = invoke("doctor")

    assert result.exit_code == 0
    verify(result.output, scrubber=make_output_scrubber())


def test_doctor_no_mocking() -> None:
    result = invoke("doctor")
    assert result.exception is None
