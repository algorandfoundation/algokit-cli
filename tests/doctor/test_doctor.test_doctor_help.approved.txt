Usage: algokit doctor [OPTIONS]

  Diagnose potential environment issues that may affect AlgoKit

  Will search the system for AlgoKit dependencies and show their versions, as
  well as identifying any potential issues.

Options:
  -c, --copy-to-clipboard  Copy the contents of the doctor message (in Markdown
                           format) in your clipboard.
  -h, --help               Show this message and exit.
